### PR TITLE
Ensure all documents have source ranges

### PR DESCRIPTION
**Pull request by rictic**
- [x] CHANGELOG.md has been updated

### CODE REVIEW

File: src/css/css-parser.ts [hunk 1]
```diff
@@ -35,7 +35,7 @@ export class CssParser implements Parser<ParsedCssDocument> {
       url,
       contents,
       ast,
-      locationOffset: inlineInfo.astNode,
+      locationOffset: inlineInfo.locationOffset,
```
Comment (rictic) [inline src/css/css-parser.ts]: `astNode` is `any` for CSS Documents right now -.-

File: src/model/inline-document.ts [hunk 2]
```diff
@@ -114,11 +114,11 @@ export function getLocationOffsetOfStartOfTextContent(node: ASTNode):
         `${util.inspect(node)}`);
   }
   if (isLocationInfo(bestLocation)) {
-    return {line: bestLocation.line - 1, col: bestLocation.col};
+    return {line: bestLocation.line - 1, col: bestLocation.col - 1};
   } else {
     return {
       line: bestLocation.startTag.line - 1,
-      col: bestLocation.startTag.endOffset,
+      col: bestLocation.startTag.endOffset - 1,
```
Comment (rictic) [inline src/model/inline-document.ts]: I missed these when I realized that parse5 ast nodes are 1 indexed in both line and column

File: src/parser/document.ts [hunk 3]
```diff
@@ -34,7 +34,16 @@ export abstract class ParsedDocument<AstNode, Visitor> {
...
+  sourceRange: SourceRange;
+
+  private readonly _locationOffset: LocationOffset|undefined;
+
+  /**
+   * The 0-based offsets into `contents` of all newline characters.
+   *
+   * Useful for converting between string offsets and SourcePositions.
+   */
+  readonly newlineIndexes: number[] = [];
```
Comment (rictic) [inline src/parser/document.ts]: Will use this more in a followup PR. Will make some source range operations in ParsedHtmlDocument much faster.

File: src/test/analyzer_test.ts [hunk 4]
```diff
@@ -683,6 +698,35 @@ var DuplicateNamespace = {};
...
+              .filter((d) => !d.isInline)
+              .map((d) => d.url)
+              .sort(),
+          [
+            'cyclic-a.html',
+            'cyclic-b.html',
+            'root.html',
+            'leaf.html',
+            'subdir/subdir-leaf.html',
+            'subdir/root-in-subdir.html'
```
Comment (rictic) [inline src/test/analyzer_test.ts]: The key thing here is that bower_components/imported.html isn't there unless you explicitly ask for it.

File: src/test/editor-service/editor-service_test.ts [hunk 5]
```diff
@@ -436,7 +436,7 @@ function editorTests(editorFactory: (basedir: string) => EditorService) {
           }]);
       assert.deepEqual(await underliner.underline(warnings), [`
 <script>var var var var var let const;</script>
-             ~`]);
+            ~`]);
```
Comment (rictic) [inline src/test/editor-service/editor-service_test.ts]: I knew something was suspicious here when I first wrote this test!

File: CHANGELOG.md [hunk 6]
```diff
@@ -11,7 +11,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
    Unreleased section, uncommenting the header as necessary.
 -->
 
-<!--## Unreleased-->
+## Unreleased
+
+* All Documents, ParsedDocuments, and ScannedDocuments now have correct SourceRanges. This also fixes a bug whereby getByKind('document') would fail to filter out package-external Documents.
```
Comment (FredKSchott) [inline CHANGELOG.md]: nit: code format `getByKind('document')`

File: src/css/css-parser.ts (hunk 1 above)
Comment (FredKSchott) [inline src/css/css-parser.ts]: ah jeez, good catch

File: src/model/document.ts [hunk 7]
```diff
@@ -42,12 +42,12 @@ export class ScannedDocument {
   warnings: Warning[];
   isInline = false;
 
-  // A scanned document has no `sourceRange` or `astNode` by default. However,
-  // they can be set later if the ScannedDocument represents an inline doc.
-  // TODO(fks): Inherit these two properties from ParsedDocument so that they do
```
Comment (FredKSchott) [inline src/model/document.ts]: ask and ye shall receive :)

File: src/parser/document.ts [hunk 8]
```diff
@@ -44,6 +53,26 @@ export abstract class ParsedDocument<AstNode, Visitor> {
     this._locationOffset = from.locationOffset;
     this.astNode = from.astNode;
     this.isInline = from.isInline;
+
+    let lastSeen = -1;
```
Comment (FredKSchott) [inline src/parser/document.ts]: `lastSeenNewline`

File: src/test/analyzer_test.ts (hunk 4 above)
Comment (FredKSchott) [inline src/test/analyzer_test.ts]: makes sense. Always good to call this out in a comment for future readers/tinkerers

**Pull request merged**